### ISSUE TITLE
`json_query` filter does not take registered variables into account

### ISSUE THREAD

**Issue by StopMotionCuber** [ansible/ansible-lint#3047]
##### Summary

With `ansible-lint 6.13.0` the following playbook incorrectly raises a `jinja[invalid]` warning even though it passes correctly when executed with `ansible`:
```yaml
- name: Grafana playbook
  hosts: localhost
  tasks:
    - name: Register previously copied dashboards
      ansible.builtin.find:
        paths: "/etc/grafana/dashboards"
        hidden: true
        patterns:
          - "*.json"
      register: dashboards_present
    - name: Get dashboard lists
      ansible.builtin.set_fact:
        _dashboards_present_list: "{{ dashboards_present | json_query('files[*].path') | default([]) }}"
```

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console
ansible-lint --version
ansible-lint 6.13.1.dev0 using ansible 2.14.2
You are using a pre-release version of ansible-lint.
```

- ansible installation method: OS package, Arch Linux
- ansible-lint installation method: OS package, Arch Linux

##### STEPS TO REPRODUCE

1. Use the above playbook and save it as `playbook.yml`
2. Run `ansible-lint playbook.yml`

##### Desired Behavior

No errors are reported

##### Actual Behavior

I'm getting the following error:
```console
WARNING  Listing 1 violation(s) that are fatal
jinja[invalid]: Error in jmespath.search in json_query filter plugin:
'dashboards_present' is undefined
playbook.yml:13 Task/Handler: Get dashboard lists

Read documentation for instructions on how to ignore specific rule violations.

              Rule Violation Summary
 count tag            profile rule associated tags
     1 jinja[invalid] basic   formatting

Failed after min profile: 1 failure(s), 0 warning(s) on 1 files.
```

##### Additional context
I have done a `git bisect run`, the error seems to be introduced by cacaa9a6

**Comment by StopMotionCuber** [comment]
Yes, I'm sure about that. We're automatically updating to the newest release of `ansible-lint` nightly within our pipelines and this error only occurred with the latest version while not changing anything from our repository. It might be that the `bisect` run was malicious if there was a fix introduced later on.

**Comment by chriscroome** [comment]
I've had a few issues with the latest version of Ansible lint related to `json_query` as well, for example [this Ansible](https://github.com/chriscroome/ansible-role-firefox/blob/main/tasks/version.yml#L47):

```yaml
    - name: "Set a fact for the JSON version variable for {{ edition.name }}"
      ansible.builtin.set_fact:
        firefox_latest: "{{ firefox_versions.json | community.general.json_query(edition.version) }}"
```

Fails with:

```
WARNING  Listing 1 violation(s) that are fatal
jinja[invalid]: Error in jmespath.search in json_query filter plugin:
'edition' is undefined
tasks/version.yml:49 Task/Handler: Set a fact for the JSON version variable for {{ edition.name }}

Read documentation for instructions on how to ignore specific rule violations.

              Rule Violation Summary               
 count tag            profile rule associated tags 
     1 jinja[invalid] basic   formatting           

Failed after min profile: 1 failure(s), 0 warning(s) on 20 files.
```

**Comment by ssbarnea** [comment]
I created a new project and I was able to get an error, but it was funny and useful, clearly not one that I would like to ignore:
```
WARNING  Listing 1 violation(s) that are fatal
jinja[invalid]: You need to install "jmespath" prior to running json_query filter
```
Basically the linter was able to spot that while I had json_query module installed, the jmespath was not installed so it would not work.

That opens a new set of questions, as linter is only able to install missing collections and not missing python dependencies. I wonder if that would not even count as a bug in community.general collection, which would be better vendoring the module.

Still, after installing I finally reproduced the posted error. I will try to add an extra condition to ignore that one as we do not have ability to read real variable values during linting, this being possible only at runtime, when it would be too late.

**Comment by StopMotionCuber** [comment]
That's not the error I'm getting though (while reproducing I also stumbled upon that, but that's easily fixable obviously).

Can you reproduce after a `pip install jmespath` on the `ansible-lint`?
In case not, could you provide a `pip freeze`, then I'd try to reproduce with your environment

**Comment by chriscroome** [comment]
The [documentation is clear](https://docs.ansible.com/ansible/latest/collections/community/general/docsite/filter_guide_selecting_json_data.html):

> You must manually install the **jmespath** dependency on the Ansible controller before using this filter.

This is also be the case for other filters in the collection, for example [jc](https://docs.ansible.com/ansible/latest/collections/community/general/docsite/filter_guide_conversions.html#converting-to-json):

> This filter needs the [jc Python library](https://pypi.org/project/jc/) installed on the controller.

**Comment by ssbarnea** [comment]
@chriscroome Computers do not perform tasks based on documentation, so writing anything there does not make it "work". AFAIK, at this moment it is not even possible to detect if installed collections do have their python requirements present. I guess we would need some extra metadata fields for exposing python dependencies.

**Comment by brucellino** [comment]
> I guess we would need some extra metadata fields for exposing python dependencies.

I came here to see this. Was hoping that we could pass "extra_pips" to the hook (or github action), similarly to what most NodeJS hooks do.

**Comment by chriscroome** [comment]
The `.pre-commit-config.yaml` I'm using with Ansible roles is:

```yaml
repos:
  - repo: https://github.com/adrienverge/yamllint.git
    rev: v1.32.0
    hooks:
      - id: yamllint
        name: YAML Lint
  - repo: https://github.com/ansible/ansible-lint.git
    rev: v6.17.2
    hooks:
      - id: ansible-lint
        name: Ansible Lint
        language: python
        additional_dependencies:
          - jc==1.23.2
          - jmespath==1.0.1
  - repo: https://github.com/jackdewinter/pymarkdown.git
    rev: v0.9.11
    hooks:
      - id: pymarkdown
        name: Markdown Lint
```
